### DISCUSSION TITLE
German translation progress

### DISCUSSION THREAD

**Discussion by nilslindemann** [fastapi/fastapi#10582] [Translations]
This is an overview over the German translation progress.

I am done with going through these documents, they are translated to German (They have PRs waiting for approval) and (currently) up-to-date, with the following exceptions:

* Docs are about to be updated:
  [sql-databases.md](https://fastapi.tiangolo.com/tutorial/sql-databases/)
  [testing-database.md](https://fastapi.tiangolo.com/advanced/testing-database/)
* Deprecated:
  [async-sql-encode-databases.md](https://fastapi.tiangolo.com/how-to/async-sql-encode-databases/)
  [nosql-databases-couchbase.md](https://fastapi.tiangolo.com/how-to/nosql-databases-couchbase/)
  [sql-databases-peewee.md](https://fastapi.tiangolo.com/how-to/sql-databases-peewee/)
* I am too lazy (3500 lines):
  [release-notes.md](https://fastapi.tiangolo.com/release-notes/)

Each of these PRs **needs two approvals** from native German speakers, before @tiangolo considers them for merging. **See the green `Review changes` button**, under the `Files changed` tab, **on the page of a PR**. [More info here](https://cca8f20c.fastapitiangolo.pages.dev/contributing/#translations) ([DE](https://cca8f20c.fastapitiangolo.pages.dev/de/contributing/#ubersetzungen)). You may have a look at [this list I created](https://github.com/tiangolo/fastapi/discussions/10369).

Thanks to @0x4Dark, @DevSpace88, @fhabers21, @JohannesJungbluth, @mawassk for translating and @creyD, @devlopersabbir, @dtleal, @DevSpace88, @fhabers21, @GeorchW, @leo-hsk, @leonlowitzki, @mattkoehne and especially @dpinezich for reviewing. :heart:

Also, a big thanks to @tiangolo, who trusted me and merged a lot of these PRs without a reviewer.

**124** of **126** PRs merged.

`index.md` [merged](https://github.com/tiangolo/fastapi/pull/10283)
`features.md` ... [merged](https://github.com/tiangolo/fastapi/pull/10284)
`learn/index.md` ... [merged](https://github.com/tiangolo/fastapi/pull/10497)
`python-types.md` ... [merged](https://github.com/tiangolo/fastapi/pull/10287)
`async.md` ... [merged](https://github.com/tiangolo/fastapi/pull/10449)
`tutorial/index.md` ... [merged](https://github.com/tiangolo/fastapi/pull/9502)
`tutorial/first-steps.md` ... [merged](https://github.com/tiangolo/fastapi/pull/9530)
`tutorial/path-params.md` ... [merged](https://github.com/tiangolo/fastapi/pull/10290)
`tutorial/query-params.md` ... [merged](https://github.com/tiangolo/fastapi/pull/10293)
`tutorial/body.md` ... [merged](https://github.com/tiangolo/fastapi/pull/10295)
`tutorial/query-params-str-validations.md` ... [merged](https://github.com/tiangolo/fastapi/pull/10304)
`tutorial/path-params-numeric-validations.md` ... [merged](https://github.com/tiangolo/fastapi/pull/10307)
`tutorial/body-multiple-params.md` ... [merged](https://github.com/tiangolo/fastapi/pull/10308)
`tutorial/body-fields.md` ... [merged](https://github.com/tiangolo/fastapi/pull/10310)
`tutorial/body-nested-models.md` ... [merged](https://github.com/tiangolo/fastapi/pull/10313)
`tutorial/schema-extra-example.md` ... [merged](https://github.com/tiangolo/fastapi/pull/10597)
`tutorial/extra-data-types.md` ... [merged](https://github.com/tiangolo/fastapi/pull/10534)
`tutorial/cookie-params.md` ... [merged](https://github.com/tiangolo/fastapi/pull/10323)
`tutorial/header-params.md` ... [merged](https://github.com/tiangolo/fastapi/pull/10326)
`tutorial/response-model.md` ... [merged](https://github.com/tiangolo/fastapi/pull/10345)
`tutorial/extra-models.md` ... [merged](https://github.com/tiangolo/fastapi/pull/10351)
`tutorial/response-status-code.md` ... [merged](https://github.com/tiangolo/fastapi/pull/10357)
`tutorial/request-forms.md` ... [merged](https://github.com/tiangolo/fastapi/pull/10361)
`tutorial/request-files.md` ... [merged](https://github.com/tiangolo/fastapi/pull/10364)
`tutorial/request-forms-and-files.md` ... [merged](https://github.com/tiangolo/fastapi/pull/10368)
`tutorial/handling-errors.md` ... [merged](https://github.com/tiangolo/fastapi/pull/10379)
`tutorial/path-operation-configuration.md` ... [merged](https://github.com/tiangolo/fastapi/pull/10383)
`tutorial/encoder.md` ... [merged](https://github.com/tiangolo/fastapi/pull/10385)
`tutorial/body-updates.md` ... [merged](https://github.com/tiangolo/fastapi/pull/10396)
`tutorial/dependencies/index.md` ... [merged](https://github.com/tiangolo/fastapi/pull/10399)
`tutorial/dependencies/classes-as-dependencies.md` ... [merged](https://github.com/tiangolo/fastapi/pull/10407)
`tutorial/dependencies/sub-dependencies.md` ... [merged](https://github.com/tiangolo/fastapi/pull/10409)
`tutorial/dependencies/dependencies-in-path-operation-decorators.md` ... [merged](https://github.com/tiangolo/fastapi/pull/10411)
`tutorial/dependencies/global-dependencies.md` ... [merged](https://github.com/tiangolo/fastapi/pull/10420)
`tutorial/dependencies/dependencies-with-yield.md` ... [merged](https://github.com/tiangolo/fastapi/pull/10422)
`tutorial/security/index.md` ... [merged](https://github.com/tiangolo/fastapi/pull/10429)
`tutorial/security/first-steps.md` ... [merged](https://github.com/tiangolo/fastapi/pull/10432)
`tutorial/security/get-current-user.md` ... [merged](https://github.com/tiangolo/fastapi/pull/10439)
`tutorial/security/simple-oauth2.md` ... [merged](https://github.com/tiangolo/fastapi/pull/10504)
`tutorial/security/oauth2-jwt.md` ... [merged](https://github.com/tiangolo/fastapi/pull/10522)
`tutorial/middleware.md` ... [merged](https://github.com/tiangolo/fastapi/pull/10391)
`tutorial/cors.md` ... [PR](https://github.com/tiangolo/fastapi/pull/10430)
`tutorial/sql-databases.md` ... ([docs are about to be updated](https://fastapi.tiangolo.com/tutorial/sql-databases/))
`tutorial/bigger-applications.md` ... [merged](https://github.com/tiangolo/fastapi/pull/10554) 
`tutorial/background-tasks.md` ... [merged](https://github.com/tiangolo/fastapi/pull/10566)
`tutorial/metadata.md` ... [merged](https://github.com/tiangolo/fastapi/pull/10581)
`tutorial/static-files.md` ... [merged](https://github.com/tiangolo/fastapi/pull/10584)
`tutorial/testing.md` ... [merged](https://github.com/tiangolo/fastapi/pull/10586)
`tutorial/debugging.md` ... [PR, depends on another PR](https://github.com/tiangolo/fastapi/pull/10591)
`advanced/index.md` ... [merged](https://github.com/tiangolo/fastapi/pull/10611)
`advanced/path-operation-advanced-configuration.md` ... [merged](https://github.com/tiangolo/fastapi/pull/10612)
`advanced/additional-status-codes.md` ... [merged](https://github.com/tiangolo/fastapi/pull/10617)
`advanced/response-directly.md` ... [merged](https://github.com/tiangolo/fastapi/pull/10618)
`advanced/custom-response.md` ... [merged](https://github.com/tiangolo/fastapi/pull/10624)
`advanced/additional-responses.md` ... [merged](https://github.com/tiangolo/fastapi/pull/10626)
`advanced/response-cookies.md` ... [merged](https://github.com/tiangolo/fastapi/pull/10627)
`advanced/response-headers.md` ... [merged](https://github.com/tiangolo/fastapi/pull/10628)
`advanced/response-change-status-code.md` ... [merged](https://github.com/tiangolo/fastapi/pull/10632)
`advanced/advanced-dependencies.md` ... [merged](https://github.com/tiangolo/fastapi/pull/10633)
`advanced/security/index.md` ... [merged](https://github.com/tiangolo/fastapi/pull/10635)
`advanced/security/oauth2-scopes.md` ... [merged](https://github.com/tiangolo/fastapi/pull/10643)
`advanced/security/http-basic-auth.md` ... [merged](https://github.com/tiangolo/fastapi/pull/10651)
`advanced/using-request-directly.md` ... [merged](https://github.com/tiangolo/fastapi/pull/10653)
`advanced/dataclasses.md` ... [merged](https://github.com/tiangolo/fastapi/pull/10667)
`advanced/middleware.md` ... [merged](https://github.com/tiangolo/fastapi/pull/10668)
`advanced/sub-applications.md` ... [merged](https://github.com/tiangolo/fastapi/pull/10671)
`advanced/behind-a-proxy.md` ... [merged](https://github.com/tiangolo/fastapi/pull/10675)
`advanced/templates.md` ... [merged](https://github.com/tiangolo/fastapi/pull/10678)
`advanced/websockets.md` ... [merged](https://github.com/tiangolo/fastapi/pull/10687)
`advanced/events.md` ... [merged](https://github.com/tiangolo/fastapi/pull/10693)
`advanced/testing-websockets.md` ... [merged](https://github.com/tiangolo/fastapi/pull/10703)
`advanced/testing-events.md` ... [merged](https://github.com/tiangolo/fastapi/pull/10704)
`advanced/testing-dependencies.md` ... [merged](https://github.com/tiangolo/fastapi/pull/10706)
`advanced/testing-database.md` ... ([docs are about to be updated](https://fastapi.tiangolo.com/advanced/testing-database/))
`advanced/async-tests.md` ... [merged](https://github.com/tiangolo/fastapi/pull/10708)
`advanced/settings.md` ... [merged](https://github.com/tiangolo/fastapi/pull/10709)
`advanced/openapi-callbacks.md` ... [merged](https://github.com/tiangolo/fastapi/pull/10710)
`advanced/openapi-webhooks.md` ... [merged](https://github.com/tiangolo/fastapi/pull/10712)
`advanced/wsgi.md` ... [merged](https://github.com/tiangolo/fastapi/pull/10713)
`advanced/generate-clients.md` ... [merged](https://github.com/tiangolo/fastapi/pull/10725)
`deployment/index.md` ... [merged](https://github.com/tiangolo/fastapi/pull/10733)
`deployment/versions.md` ... [merged](https://github.com/tiangolo/fastapi/pull/10491)
`deployment/https.md` ... [merged](https://github.com/tiangolo/fastapi/pull/10737)
`deployment/manually.md` ... [merged](https://github.com/tiangolo/fastapi/pull/10738)
`deployment/concepts.md` ... [merged](https://github.com/tiangolo/fastapi/pull/10744)
`deployment/cloud.md` ... [merged](https://github.com/tiangolo/fastapi/pull/10746)
`deployment/server-workers.md` ... [merged](https://github.com/tiangolo/fastapi/pull/10747)
`deployment/docker.md` ... [merged](https://github.com/tiangolo/fastapi/pull/10759)
`how-to/index.md` ... [merged](https://github.com/tiangolo/fastapi/pull/10769)
`how-to/general.md` ... [merged](https://github.com/tiangolo/fastapi/pull/10770)
`how-to/sql-databases-peewee.md` ... ([deprecated](https://fastapi.tiangolo.com/how-to/sql-databases-peewee/))
`how-to/async-sql-encode-databases.md` ... ([deprecated](https://fastapi.tiangolo.com/how-to/async-sql-encode-databases/))
`how-to/nosql-databases-couchbase.md` ... ([deprecated](https://fastapi.tiangolo.com/how-to/nosql-databases-couchbase/))
`how-to/graphql.md` ... [merged](https://github.com/tiangolo/fastapi/pull/10788)
`how-to/custom-request-and-route.md` ... [merged](https://github.com/tiangolo/fastapi/pull/10789)
`how-to/conditional-openapi.md` ... [merged](https://github.com/tiangolo/fastapi/pull/10790)
`how-to/extending-openapi.md` ... [merged](https://github.com/tiangolo/fastapi/pull/10794)
`how-to/separate-openapi-schemas.md` ... [merged](https://github.com/tiangolo/fastapi/pull/10796)
`how-to/custom-docs-ui-assets.md` ... [merged](https://github.com/tiangolo/fastapi/pull/10803)
`how-to/configure-swagger-ui.md` ... [merged](https://github.com/tiangolo/fastapi/pull/10804)
`reference/index.md` ... [merged](https://github.com/tiangolo/fastapi/pull/10497)
`reference/fastapi.md` ... [merged](https://github.com/tiangolo/fastapi/pull/10813)
`reference/parameters.md` ... [merged](https://github.com/tiangolo/fastapi/pull/10814)
`reference/status.md` ... [merged](https://github.com/tiangolo/fastapi/pull/10815)
`reference/uploadfile.md` ... [merged](https://github.com/tiangolo/fastapi/pull/10816)
`reference/exceptions.md` ... [merged](https://github.com/tiangolo/fastapi/pull/10817)
`reference/dependencies.md` ... [merged](https://github.com/tiangolo/fastapi/pull/10818)
`reference/apirouter.md` ... [merged](https://github.com/tiangolo/fastapi/pull/10819)
`reference/background.md` ... [merged](https://github.com/tiangolo/fastapi/pull/10820)
`reference/request.md` ... [merged](https://github.com/tiangolo/fastapi/pull/10821)
`reference/websockets.md` ... [merged](https://github.com/tiangolo/fastapi/pull/10822)
`reference/httpconnection.md` ... [merged](https://github.com/tiangolo/fastapi/pull/10823)
`reference/response.md` ... [merged](https://github.com/tiangolo/fastapi/pull/10824)
`reference/responses.md` ... [merged](https://github.com/tiangolo/fastapi/pull/10825)
`reference/middleware.md` ... [merged](https://github.com/tiangolo/fastapi/pull/10837)
`reference/openapi/index.md` ... [merged](https://github.com/tiangolo/fastapi/pull/10838)
`reference/openapi/docs.md` ... [merged](https://github.com/tiangolo/fastapi/pull/10838)
`reference/openapi/models.md` ... [merged](https://github.com/tiangolo/fastapi/pull/10838)
`reference/security/index.md` ... [merged](https://github.com/tiangolo/fastapi/pull/10839)
`reference/encoders.md` ... [merged](https://github.com/tiangolo/fastapi/pull/10840)
`reference/staticfiles.md` ... [merged](https://github.com/tiangolo/fastapi/pull/10841)
`reference/templating.md` ... [merged](https://github.com/tiangolo/fastapi/pull/10842)
`reference/testclient.md` ... [merged](https://github.com/tiangolo/fastapi/pull/10843)
`fastapi-people.md` ... [merged](https://github.com/tiangolo/fastapi/pull/10285)
`resources/index.md` ... [merged](https://github.com/tiangolo/fastapi/pull/10497)
`project-generation.md` ... [merged](https://github.com/tiangolo/fastapi/pull/10851)
`external-links.md` ... [merged](https://github.com/tiangolo/fastapi/pull/10852)
`newsletter.md` ... [merged](https://github.com/tiangolo/fastapi/pull/10853)
`about/index.md` ... [merged](https://github.com/tiangolo/fastapi/pull/10497)
`alternatives.md` ... [merged](https://github.com/tiangolo/fastapi/pull/10855)
`history-design-future.md` ... [merged](https://github.com/tiangolo/fastapi/pull/10865)
`benchmarks.md` ... [merged](https://github.com/tiangolo/fastapi/pull/10866)
`help/index.md` ... [merged](https://github.com/tiangolo/fastapi/pull/10497)
`help-fastapi.md` ... [merged](https://github.com/tiangolo/fastapi/pull/10455)
`contributing.md` ... [merged](https://github.com/tiangolo/fastapi/pull/10487)
`release-notes.md` ... (too much work, 3500 lines and will change often)

**Comment by T-256** [comment]
@tiangolo could open a branch for you to push your change on that branch (e.g. `new-translate/German`).
After all files translated, it can be pushed to main branch via a PR.

**Reply by nilslindemann** [reply]
@T-256 What is the advantage of doing so? (I hope that's not a stupid question ^^)

**Reply by T-256** [reply]
No duplicated commits into main branch.
 
Less opened PRs (if the maintainer merge instantly), so it'd be easier to search for codebase changes in _Pull Requests_ section

**Reply by nilslindemann** [reply]
I am open to this. On the other hand, I am anyway nearly done (six documents left) so it would not really make a difference, would it?

**Reply by T-256** [reply]
> On the other hand, I am anyway nearly done (six documents left)

Then that's fine, go ahead.
IIUC, at end only https://github.com/tiangolo/fastapi/pull/10625 should be merged to main branch?

**Reply by nilslindemann** [reply]
#10625 currently is just there to provide a preview of the state of translation, to which I link [above](https://github.com/tiangolo/fastapi/discussions/10582#discussion-5810217). My intent is to delete that PR, when all the other related smaller per-document-PRs (see the numbers behind the commit titles) are merged. It also contains some changes which are not yet a PR (commits without a number after them), or which are not intended to be merged (commits starting with "(Local)"). So, in its current state, it should not be merged, I would have to undo a few commits.

On the other hand, as you mention it: It would be easier for me, if these were all merged as one big PR and people can create their own PRs afterwards, improving things. From January 23, 2024 on, I will have a training as an IT specialist in Bremen for the next two years, and then I will have less time.

**Reply by T-256** [reply]
I think main branch will be crowded by merging about 120 PRs for just translated docs of a new language.
As I said above you can:
1. ask maintainer to open new branch
2. rebase all of PR's to that created branch
3. ask maintainer to review and merge PR's

when all merged, maintainer can push full documentation behind a single PR from new branch

**Reply by nilslindemann** [reply]
Oh, ok, it sounded to me as if you are one of the core devs, and you want to do it in a specific way. Otherwise, I do not think that Tiangolo wants to change the procedure, which consists of

* [one PR per document](https://fastapi.tiangolo.com/contributing/#translations:~:text=Add%20a%20single%20pull%20request%20per%20page%20translated) ("Add a single pull request per page translated"), and
* [two approvals per PR by native speakers](https://fastapi.tiangolo.com/contributing/#translations:~:text=I%27ll%20wait%20for%20several%20others%20to%20review%20the%20translation) ("For the languages I don't speak, I'll wait for several others to review the translation before merging")

**Reply by T-256** [reply]
Uh, I think it could be better, but it's what it is now, it seems there is no problem for you work.

@tiangolo you can check my proposed suggestion for document PRs, it could be less crowded on `main` branch.

**Reply by nilslindemann** [reply]
For now, you can also filter the list, for example:

* [is:open is:pr -author:nilslindemann](https://github.com/tiangolo/fastapi/pulls?q=is%3Aopen+is%3Apr+-author%3Anilslindemann) hides my PRs
* [is:open is:pr NOT "translation" in:title](https://github.com/tiangolo/fastapi/pulls?q=is%3Aopen+is%3Apr+NOT+%22translation%22+in%3Atitle) hides all PRs with "translation" in the title
